File: clients/java/src/test/java/io/volant/ReassignPartitionsTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** ReassignPartitions client tests against a scripted TCP broker (no live server). */
class ReassignPartitionsTest {
    @Test
    void successReturnsGeneration() throws Exception {
        try (ReassignPartitionsServer srv = ReassignPartitionsServer.ok(7)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                int got = c.reassignPartitions("events", Integer.valueOf(0), new int[] {1, 2});
                assertEquals(7, got);
            }
            assertEquals("events", srv.topic.get());
            assertEquals(0L, srv.partition.get());
            assertEquals(List.of(1L, 2L), srv.replicas.get());
        }
    }

    @Test
    void nullPartitionEncodesAllSentinel() throws Exception {
        try (ReassignPartitionsServer srv = ReassignPartitionsServer.ok(3)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                int got = c.reassignPartitions("events");
                assertEquals(3, got);
            }
            assertEquals("events", srv.topic.get());
            assertEquals(Codec.REASSIGN_ALL_PARTITIONS, srv.partition.get());
            assertTrue(srv.replicas.get().isEmpty());
        }
    }

    @Test
    void nonzeroErrorCodeRaises() throws Exception {
        try (ReassignPartitionsServer srv = ReassignPartitionsServer.error(2)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(
                        BrokerException.class,
                        () -> c.reassignPartitions("missing", new int[] {1, 2}));
                assertEquals(2, ex.code);
                assertEquals("reassign_partitions", ex.op);
            }
        }
    }

    private static final class ReassignPartitionsServer implements AutoCloseable {
        final int port;
        final AtomicReference<String> topic = new AtomicReference<>();
        final AtomicLong partition = new AtomicLong();
        final AtomicReference<List<Long>> replicas = new AtomicReference<>(List.of());
        private final int errorCode;
        private final long generation;
        private final ServerSocket listen;
        private final Thread thread;
        private final AtomicReference<Exception> error = new AtomicReference<>();

        static ReassignPartitionsServer ok(long generation) throws IOException {
            return new ReassignPartitionsServer(0, generation);
        }

        static ReassignPartitionsServer error(int code) throws IOException {
            return new ReassignPartitionsServer(code, 0);
        }

        private ReassignPartitionsServer(int errorCode, long generation) throws IOException {
            this.errorCode = errorCode;
            this.generation = generation;
            listen = new ServerSocket(0, 8, InetAddress.getByName("127.0.0.1"));
            listen.setSoTimeout(8_000);
            port = listen.getLocalPort();
            thread = new Thread(this::serve, "volant-reassign-partitions");
            thread.setDaemon(true);
            thread.start();
        }

        private void serve() {
            try (Socket conn = listen.accept()) {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    if (d.frame.opcode != Codec.OP_REASSIGN_PARTITIONS) {
                        error.set(new ProtocolException("unexpected opcode " + d.frame.opcode));
                        return;
                    }
                    Codec.ReassignPartitionsRequest req =
                            Codec.decodeReassignPartitionsRequest(d.frame.payload);
                    topic.set(req.topic);
                    partition.set(req.partition);
                    replicas.set(new ArrayList<>(req.replicas));
                    long gen = errorCode != 0 ? 0 : generation;
                    byte[] payload = Codec.encodeReassignPartitionsResponse(
                            new Codec.ReassignPartitionsResponse(errorCode, gen));
                    out.write(
                            Frame.encode(
                                    Codec.OP_REASSIGN_PARTITIONS_RESPONSE,
                                    d.frame.correlationId,
                                    payload));
                    out.flush();
                }
            } catch (Exception e) {
                error.set(e);
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: clients/go/reassign_partitions_test.go
package volant_test

import (
	"errors"
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

type reassignPartitionsServerResult struct {
	topic     string
	partition uint32
	replicas  []uint32
	err       error
}

func serveReassignPartitions(t *testing.T, errorCode uint16, generation uint32) (addr string, got *reassignPartitionsServerResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &reassignPartitionsServerResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			if f.Opcode != codec.OpReassignPartitions {
				res.err = &frame.ProtocolError{Msg: "unexpected opcode"}
				return
			}
			req, err := codec.DecodeReassignPartitionsRequest(f.Payload)
			if err != nil {
				res.err = err
				return
			}
			res.topic = req.Topic
			res.partition = req.Partition
			res.replicas = req.Replicas
			gen := generation
			if errorCode != 0 {
				gen = 0
			}
			payload, err := codec.EncodeReassignPartitionsResponse(codec.ReassignPartitionsResponse{
				ErrorCode:  errorCode,
				Generation: gen,
			})
			if err != nil {
				res.err = err
				return
			}
			raw, err := frame.Encode(codec.OpReassignPartitionsResponse, f.CorrelationID, payload)
			if err != nil {
				res.err = err
				return
			}
			if _, err := conn.Write(raw); err != nil {
				res.err = err
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestReassignPartitionsSuccessReturnsGeneration(t *testing.T) {
	addr, got, stop := serveReassignPartitions(t, 0, 7)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	part := uint32(0)
	out, err := c.ReassignPartitions("events", []uint32{1, 2}, &part)
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.topic != "events" || got.partition != 0 || len(got.replicas) != 2 || got.replicas[0] != 1 || got.replicas[1] != 2 {
		t.Fatalf("wire request topic=%q partition=%d replicas=%v", got.topic, got.partition, got.replicas)
	}
	if out != 7 {
		t.Fatalf("parsed %d", out)
	}
}

func TestReassignPartitionsNilPartitionEncodesAll(t *testing.T) {
	addr, got, stop := serveReassignPartitions(t, 0, 3)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.ReassignPartitions("events", nil, nil)
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.topic != "events" || got.partition != codec.ReassignAllPartitions || len(got.replicas) != 0 {
		t.Fatalf("wire request topic=%q partition=%d replicas=%v", got.topic, got.partition, got.replicas)
	}
	if out != 3 {
		t.Fatalf("parsed %d", out)
	}
}

func TestReassignAllPartitionsEncodesAll(t *testing.T) {
	addr, got, stop := serveReassignPartitions(t, 0, 3)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.ReassignAllPartitions("events", []uint32{1, 2})
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.topic != "events" || got.partition != codec.ReassignAllPartitions || len(got.replicas) != 2 || got.replicas[0] != 1 || got.replicas[1] != 2 {
		t.Fatalf("wire request topic=%q partition=%d replicas=%v", got.topic, got.partition, got.replicas)
	}
	if out != 3 {
		t.Fatalf("parsed %d", out)
	}
}

func TestReassignPartitionsNonzeroErrorRaises(t *testing.T) {
	addr, _, stop := serveReassignPartitions(t, 2, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.ReassignPartitions("missing", []uint32{1, 2}, nil)
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 2 || be.Op != "reassign_partitions" {
		t.Fatalf("broker error %+v", be)
	}
}
